src/main/webapp/go.js

```diff
@@ -299,19 +340,39 @@ function sendIndexToServlet(index) {
  * Disable lint check because proceed() is called from go.html.
  */
 function proceed() { //eslint-disable-line
-  destIndex++;
-  sendIndexToServlet(destIndex);
-  if (destIndex < huntArr.length) {
-    updateMessage(RIDDLE_DISPLAY, 'Riddle: ' + huntArr[destIndex].puzzle);
-    toggleHintButton(/* hide = */ false);
+  hunt.destIndex = hunt.destIndex + 1;
+  sendIndexToServlet(hunt.getDestIndex());
+  if (hunt.getDestIndex() < hunt.getNumItems()) {
+    updateMessage(RIDDLE_DISPLAY, 'Riddle: ' + hunt.getCurDestPuzzle());
+    delayHintButton();
   } else {
     updateRiddleToFinalMessage();
     toggleHintButton(/* hide = */ true);
   }
+  updateProgressBar();
   toggleProceedButton(/* hide = */ true);
   deleteMessage(SUBMIT_DISPLAY);
   deleteMessage(HINT_DISPLAY);
-  hintIndex = 0;
+  hunt.setHintIndex(0);
+}
+
+/**
+ * Update the progress bar to reflect how far the user is
+ * on the scavenger hunt.
+ */
+function updateProgressBar() {
+  const bar = document.getElementById(PROGRESS_DISPLAY);
+  bar.style.width = hunt.getProgress() + '%';
+}
+
+/**
+ * When the user first begins to find a destination, hide the hint
+ * button. Only show the hint button after a certain amount of time
+ * has passed.
+ */
+function delayHintButton() {
+  toggleHintButton(/* hide = */ true);
+  window.setTimeout(toggleHintButton, HIDE_INTERVAL_MS, false);
```

Comment:
Does this need to be disabled once the user finds the current destination? Probably fine for now with a 5 second delay, but I think this should be a TODO in order to more cleanly handle the possibility of having longer delays. I imagine this situation could happen (let's pretend it's a 5 minute delay)
User starts scavenger hunt at t=0 minutes
User gets correct first destination and proceeds to next at t=4 minutes
User is trying to figure out the new destination, and at t=5 minutes, they get access to the hint button even though they have been looking for the current destination for only 1 minute.